File: regions/system/library/dart/auth/test/pkce_test.dart
import 'dart:math';

import 'package:k1s0_auth/auth.dart';
import 'package:test/test.dart';

void main() {
  group('PKCE', () {
    group('generateCodeVerifier', () {
      test('長さ 43 の文字列を生成すること', () {
        final verifier = generateCodeVerifier();
        expect(verifier.length, equals(43));
      });

      test('未予約文字のみを含むこと', () {
        final verifier = generateCodeVerifier();
        // RFC 7636: unreserved characters [A-Z] / [a-z] / [0-9] / "-" / "." / "_" / "~"
        expect(
          verifier,
          matches(RegExp(r'^[A-Za-z0-9\-._~]+$')),
        );
      });

      test('呼び出しごとにユニークな値を生成すること', () {
        final v1 = generateCodeVerifier();
        final v2 = generateCodeVerifier();
        expect(v1, isNot(equals(v2)));
      });

      test('注入された乱数生成器を使用すること', () {
        // Fixed seed for deterministic output
        final rng = Random(42);
        final v1 = generateCodeVerifier(random: Random(42));
        final v2 = generateCodeVerifier(random: Random(42));
        expect(v1, equals(v2));

        // Different seed
        final v3 = generateCodeVerifier(random: rng);
        // After consuming from rng, a new Random(42) should differ
        final v4 = generateCodeVerifier(random: rng);
        expect(v3, isNot(equals(v4)));
      });
    });

    group('generateCodeChallenge', () {
      test('base64url エンコードされた SHA-256 ハッシュを生成すること', () {
        final verifier = 'dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk';
        final challenge = generateCodeChallenge(verifier);
        // base64url: [A-Za-z0-9_-] only, no padding
        expect(challenge, matches(RegExp(r'^[A-Za-z0-9_-]+$')));
      });

      test('同じ入力に対して一貫した出力を生成すること', () {
        const verifier = 'test-verifier-value';
        final c1 = generateCodeChallenge(verifier);
        final c2 = generateCodeChallenge(verifier);
        expect(c1, equals(c2));
      });

      test('異なる入力に対して異なる出力を生成すること', () {
        final c1 = generateCodeChallenge('verifier-1');
        final c2 = generateCodeChallenge('verifier-2');
        expect(c1, isNot(equals(c2)));
      });

      test('パディング文字を含まないこと', () {
        for (var i = 0; i < 10; i++) {
          final verifier = generateCodeVerifier();
          final challenge = generateCodeChallenge(verifier);
          expect(challenge, isNot(contains('=')));
        }
      });

      test('+ または / 文字を含まないこと', () {
        for (var i = 0; i < 10; i++) {
          final verifier = generateCodeVerifier();
          final challenge = generateCodeChallenge(verifier);
          expect(challenge, isNot(contains('+')));
          expect(challenge, isNot(contains('/')));
        }
      });
    });

    group('base64UrlEncode', () {
      test('空リストをエンコードできること', () {
        expect(base64UrlEncode([]), equals(''));
      });

      test('パディングなしでエンコードすること', () {
        final encoded = base64UrlEncode([1]);
        expect(encoded, isNot(contains('=')));
      });

      test('URL セーフな文字を使用すること', () {
        // Values known to produce + and / in standard base64
        final encoded = base64UrlEncode([251, 255, 254]);
        expect(encoded, isNot(contains('+')));
        expect(encoded, isNot(contains('/')));
      });
    });
  });
}

File: regions/system/library/typescript/auth/__tests__/pkce.test.ts
import { describe, it, expect } from 'vitest';
import {
  generateCodeVerifier,
  generateCodeChallenge,
  base64UrlEncode,
} from '../src/pkce';

describe('PKCE', () => {
  describe('generateCodeVerifier', () => {
    it('should generate a base64url-encoded string', () => {
      const verifier = generateCodeVerifier();
      // Base64url: only contains [A-Za-z0-9_-]
      expect(verifier).toMatch(/^[A-Za-z0-9_-]+$/);
    });

    it('should generate a string of appropriate length', () => {
      const verifier = generateCodeVerifier();
      // 32 bytes -> ~43 chars in base64url (no padding)
      expect(verifier.length).toBeGreaterThanOrEqual(40);
      expect(verifier.length).toBeLessThanOrEqual(48);
    });

    it('should generate unique values on each call', () => {
      const v1 = generateCodeVerifier();
      const v2 = generateCodeVerifier();
      expect(v1).not.toBe(v2);
    });

    it('should use the injected random function', () => {
      const mockGetRandomValues = (array: Uint8Array): Uint8Array => {
        for (let i = 0; i < array.length; i++) {
          array[i] = i;
        }
        return array;
      };
      const v1 = generateCodeVerifier(mockGetRandomValues);
      const v2 = generateCodeVerifier(mockGetRandomValues);
      expect(v1).toBe(v2);
    });

    it('should not contain +, /, or = characters', () => {
      for (let i = 0; i < 10; i++) {
        const verifier = generateCodeVerifier();
        expect(verifier).not.toContain('+');
        expect(verifier).not.toContain('/');
        expect(verifier).not.toContain('=');
      }
    });
  });

  describe('generateCodeChallenge', () => {
    it('should generate a base64url-encoded SHA-256 hash', async () => {
      const verifier = 'dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk';
      const challenge = await generateCodeChallenge(verifier);
      // base64url encoded
      expect(challenge).toMatch(/^[A-Za-z0-9_-]+$/);
    });

    it('should produce consistent output for the same input', async () => {
      const verifier = 'test-verifier-value';
      const c1 = await generateCodeChallenge(verifier);
      const c2 = await generateCodeChallenge(verifier);
      expect(c1).toBe(c2);
    });

    it('should produce different output for different inputs', async () => {
      const c1 = await generateCodeChallenge('verifier-1');
      const c2 = await generateCodeChallenge('verifier-2');
      expect(c1).not.toBe(c2);
    });

    it('should not contain +, /, or = characters', async () => {
      for (let i = 0; i < 10; i++) {
        const verifier = generateCodeVerifier();
        const challenge = await generateCodeChallenge(verifier);
        expect(challenge).not.toContain('+');
        expect(challenge).not.toContain('/');
        expect(challenge).not.toContain('=');
      }
    });

    it('should use the injected digest function', async () => {
      const mockDigest = async (
        _algorithm: string,
        _data: BufferSource,
      ): Promise<ArrayBuffer> => {
        return new Uint8Array([1, 2, 3, 4]).buffer;
      };
      const challenge = await generateCodeChallenge('test', mockDigest);
      expect(challenge).toBe(base64UrlEncode(new Uint8Array([1, 2, 3, 4])));
    });
  });

  describe('base64UrlEncode', () => {
    it('should encode an empty buffer', () => {
      expect(base64UrlEncode(new Uint8Array([]))).toBe('');
    });

    it('should encode without padding', () => {
      const encoded = base64UrlEncode(new Uint8Array([1]));
      expect(encoded).not.toContain('=');
    });

    it('should replace + with - and / with _', () => {
      // Test with values known to produce + and / in standard base64
      const encoded = base64UrlEncode(new Uint8Array([251, 255, 254]));
      expect(encoded).not.toContain('+');
      expect(encoded).not.toContain('/');
    });
  });
});
